### PR TITLE
[Edit Post]: Migrate store actions to thunks

### PR DESCRIPTION
**Pull request by ntsekouras**
This PR migrates the `edit-post` data store to thunks and updates unit tests.

### Notes

There are some tests missing right now, but when I started writing them a lot of mocking for `core` store internals were needed and I'm not sure if it's the right approach here. Maybe some e2e tests would be better for these cases?

Namely the untested actions are `__unstableCreateTemplate, initializeMetaBoxes and requestMetaBoxUpdates`.

### CODE REVIEW

File: packages/edit-post/src/store/test/actions.js [hunk 1]
```diff
@@ -1,120 +1,130 @@
...
+	// Create a registry and register used stores.
+	const registry = createRegistry();
+	[
+		editPostStore,
+		noticesStore,
+		blockEditorStore,
+		coreStore,
+		interfaceStore,
+		editorStore,
+	].forEach( registry.register );
```
Comment (ntsekouras) [inline packages/edit-post/src/store/test/actions.js:26]: Depending on my notes about any more needed tests, I'll remove any extra store registered here.

File: packages/edit-post/src/store/actions.js [hunk 2]
```diff
@@ -157,53 +146,43 @@ export function removeEditorPanel( panelName ) {
...
-	yield controls.dispatch(
-		interfaceStore,
-		isPinned ? 'unpinItem' : 'pinItem',
-		'core/edit-post',
-		pluginName
-	);
-}
+	registry
+		.dispatch( interfaceStore )
+		[ isPinned ? 'unpinItem' : 'pinItem' ]( 'core/edit-post', pluginName );
```
Comment (adamziel) [inline packages/edit-post/src/store/actions.js:184]: I had to stop here for a moment to consider all the logic. Perhaps it would read smoother without the ternary operator?

```suggestion
	const { unpinItem, pinItem} = registry.dispatch( interfaceStore );
	if ( isPinned ) {
		unpinItem( 'core/edit-post', pluginName );
	} else {
		pinItem( 'core/edit-post', pluginName );
	}
```

File: packages/edit-post/src/store/actions.js [hunk 3]
```diff
@@ -436,93 +413,80 @@ export function setIsEditingTemplate( value ) {
...
 	}
-
-	const postType = yield controls.select( editorStore, 'getCurrentPostType' );
-
 	// Only initialize once.
 	if ( metaBoxesInitialized ) {
 		return;
 	}
-
+	const postType = registry.select( editorStore ).getCurrentPostType();
```
Comment (adamziel) [inline packages/edit-post/src/store/actions.js:473]: Good call moving this lower

File: packages/edit-post/src/store/test/actions.js [hunk 4]
```diff
@@ -1,120 +1,130 @@
...
-		} );
+	it( 'togglePinnedPluginItem', () => {
+		registry.dispatch( editPostStore ).togglePinnedPluginItem( 'rigatoni' );
+		// Sidebars are pinned by default.
+		// @See https://github.com/WordPress/gutenberg/pull/21645
+		expect(
+			registry
+				.select( interfaceStore )
+				.isItemPinned( editPostStore.name, 'rigatoni' )
+		).toBeFalsy();
```
Comment (adamziel) [inline packages/edit-post/src/store/test/actions.js]: `toBeFalse()` ?

File: packages/edit-post/src/store/test/actions.js [hunk 5]
```diff
@@ -1,120 +1,130 @@
...
+		registry.dispatch( editPostStore ).togglePinnedPluginItem( 'rigatoni' );
+		expect(
+			registry
+				.select( interfaceStore )
+				.isItemPinned( editPostStore.name, 'rigatoni' )
+		).toBeTruthy();
```
Comment (adamziel) [inline packages/edit-post/src/store/test/actions.js]: `toBeTrue()` ?

File: packages/edit-post/src/store/actions.js [hunk 6]
```diff
@@ -270,25 +249,26 @@ export function showBlockTypes( blockNames ) {
...
 		metaBoxesPerLocation,
-	};
-}
+	} );
 
 /**
- * Returns an action object used to request meta box update.
- *
- * @yield {Object} Action object.
+ * Returns a promise of an action object used to request meta box update.
```
Comment (adamziel) [inline packages/edit-post/src/store/actions.js]: This should do it:

```suggestion
 * Requests a metabox update
```

Otherwise we end up writing C++ in plain english :-) _Promise of an action that requests ..._ sounds like a lot of pointers and it's hard to unwind in my head. This action is used to request an update, let's say it directly.

**Comment by adamziel** [comment]
> There are some tests missing right now, but when I started writing them a lot of mocking for core store internals were needed and I'm not sure if it's the right approach here. Maybe some e2e tests would be better for these cases?

@ntsekouras There was a similar problem in `edit-navigation`, and the solution was to use the actual store instead of mocking it:

https://github.com/WordPress/gutenberg/blob/8b1d86e250fd8e314cfa6c2dad47a35341460f4c/packages/edit-navigation/src/store/test/resolvers.js#L40-L45

It still required mocking the API response though:

https://github.com/WordPress/gutenberg/blob/8b1d86e250fd8e314cfa6c2dad47a35341460f4c/packages/edit-navigation/src/store/test/resolvers.js#L121

Would that make it easier here? If not, I think leaning on e2e is reasonable.

File: packages/edit-post/src/store/actions.js (hunk 2 above)
Comment (jsnajdr) [reply packages/edit-post/src/store/actions.js:184]: Ideally (out of scope for this PR) there would be a `togglePinnedItem` action in the `interface` store. The `pinItem`/`unpinItem` actions are used at two places, and both of them really want to toggle.

File: packages/edit-post/src/store/actions.js [hunk 7]
```diff
@@ -436,93 +413,80 @@ export function setIsEditingTemplate( value ) {
...
 	if ( ! isEditorReady ) {
 		return;
 	}
-
-	const postType = yield controls.select( editorStore, 'getCurrentPostType' );
-
 	// Only initialize once.
 	if ( metaBoxesInitialized ) {
 		return;
 	}
```
Comment (jsnajdr) [inline packages/edit-post/src/store/actions.js:472]: Interestingly, the "is editor ready" and "are metaboxes initialized" checks are already performed by the `<MetaBoxes>` component before calling `initializeMetaBoxes`. One of the checks is redundant.

File: packages/edit-post/src/store/actions.js [hunk 8]
```diff
@@ -436,93 +413,80 @@ export function setIsEditingTemplate( value ) {
...
-	const hasMetaBoxes = yield controls.select( editPostStore, 'hasMetaBoxes' );
+	let wasSavingPost = registry.select( editorStore ).isSavingPost();
+	let wasAutosavingPost = registry.select( editorStore ).isAutosavingPost();
+	const hasMetaBoxes = select.hasMetaBoxes();
 
 	// Save metaboxes when performing a full save on the post.
-	subscribe( () => {
-		const isSavingPost = select( editorStore ).isSavingPost();
-		const isAutosavingPost = select( editorStore ).isAutosavingPost();
+	registry.subscribe( async () => {
```
Comment (jsnajdr) [inline packages/edit-post/src/store/actions.js:485]: This subscription doesn't really belong to a Redux action. It's something that would be much more easily implemented in the `<MetaBoxes>` component. The component would then (elegantly) do three things:
```js
function MetaBoxes() {
  // initialize metaboxes
  useEffect( () => {
    if ( shouldInit ) { initializeMetaBoxes(); }
  }, [ shouldInit ] );

  // update metaboxes (new)
  useEffect( () => {
    if ( shouldUpdate ) { requestMetaBoxUpdates(); }
  }, [ shouldUpdate ] );

  // render metaboxes
  return <>{ metaBoxes.map( box => ... ) }</>;
}
```

File: packages/edit-post/src/store/actions.js (hunk 6 above)
Comment (jsnajdr) [reply packages/edit-post/src/store/actions.js]: Or, with even less indirection, "do a metaboxes update" 🙂 When doing thunk migrations, I often rewrite action descriptions like this. Instead of saying like "returns an action used to signal that X is requested," just say "do X."

File: packages/edit-post/src/store/actions.js [hunk 9]
```diff
@@ -338,17 +315,17 @@ export function* requestMetaBoxUpdates() {
...
 			url: window._wpMetaBoxUrl,
 			method: 'POST',
 			body: formData,
 			parse: false,
 		} );
-		yield controls.dispatch( editPostStore, 'metaBoxUpdatesSuccess' );
+		dispatch.metaBoxUpdatesSuccess();
 	} catch {
-		yield controls.dispatch( editPostStore, 'metaBoxUpdatesFailure' );
+		dispatch.metaBoxUpdatesFailure();
```
Comment (jsnajdr) [inline packages/edit-post/src/store/actions.js:326]: Interesting that both these actions do the same thing: update the `isSavingMetaBoxes` flag. They could be merged into one. We are ignoring the errors.

Also, do they need to be a part of the public API? There is no use case for that. Just like `REQUEST_META_BOX_UPDATES`, they could be internal/private.

File: packages/edit-post/src/store/test/actions.js [hunk 10]
```diff
@@ -1,120 +1,130 @@
...
-			expect( togglePublishSidebar() ).toEqual( {
-				type: 'TOGGLE_PUBLISH_SIDEBAR',
-			} );
-		} );
+		registry.dispatch( editPostStore ).toggleFeature( 'welcomeGuide' );
+		expect(
+			registry
+				.select( interfaceStore )
+				.isFeatureActive( editPostStore.name, 'welcomeGuide' )
+		).toBeFalsy();
```
Comment (jsnajdr) [inline packages/edit-post/src/store/test/actions.js]: `isFeatureActive` always returns a true boolean, not just a falsy value. A stronger `.toBe( false )` expectation is in place.

**Review by jsnajdr — APPROVED** [review]
Looks good. I posted a few comments, but they are more like refactoring ideas rather than blockers.

There was an incorrect update of `package-lock.json` and I added a commit that fixes that.

I also rebased the PR onto the latest `trunk`. Let's see what the CI checks say now.

**Comment by jsnajdr** [comment]
The e2es are failing at:
```
WP Editor Meta Boxes › Should save the changes

    Node is either not clickable or not an HTMLElement
```
The PR is changing exactly this metabox-saving code, so apparently at some point the migration went wrong.

File: packages/edit-post/src/store/actions.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-post/src/store/actions.js]: 👍

**Comment by ntsekouras** [comment]
Thanks for the reviews here! Also thank you @jsnajdr for the extra work with rebasing and fixes. I'm now looking at the comments and will try to find out about the metaboxes test failures. As I'm not really familiar with metaboxes, I could use some help here if you find something obvious, but either way I'll try to solve them.

I can see that some comments about refactoring some things in metaboxes make sense, but I'd love if we could just refactor to use thunks here and handle these in a follow up.

**Comment by jsnajdr** [comment]
> I'm now looking at the comments and will try to find out about the metaboxes test failures.

@ntsekouras I did some debugging and it seems that the e2e test is inherently fragile and flaky. Here's approximately what it does:
```js
await savePost();
reloadPage();
verifySavedInfoHasBeenLoaded();
```
I.e., it reloads the page immediately after `savePost` is finished and assumes that saving metaboxes is finished by that time, too. But is it really finished? The code where metaboxes are saved is a store listener and looks like:
```js
subscribe( async () => {
  if ( prevIsSaving === true && nextIsSaving === false ) {
    await dispatch( requestMetaBoxUpdates() );
  }
} );
```
Is the metabox save request performed before the `await savePost()` finishes, or before? That depends on exact timing of promise resolutions etc and it seems that this timing has changed in this PR. It's harmless for the real user, but the e2e test depends on it.

**Comment by ntsekouras** [comment]
Thanks for helping @jsnajdr!!

I'm still looking into the failures and it might be something related with the `initializeMetaBoxes`. I've copied pasted the code [from the plugin](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/plugins/wp-editor-metabox.php) and in this PR the metabox is not showing at all - thus the `node not found` error. I haven't pinpointed the reason for that yet.

**Comment by jsnajdr** [comment]
> I've copied pasted the code from the plugin and in this PR the metabox is not showing at all

I also tried the patch with the `wp-editor-metabox` plugin installed and all the metabox elements are there in the DOM, but they are hidden with CSS. Not sure yet whether that's expected or not.

File: packages/edit-post/src/store/actions.js [hunk 11]
```diff
@@ -509,7 +509,7 @@ export const initializeMetaBoxes = () => ( { registry, select, dispatch } ) => {
 		}
 	} );
 
-	return {
+	return dispatch( {
```
Comment (jsnajdr) [inline packages/edit-post/src/store/actions.js]: The `return` should be removed as it doesn't do anything useful, but yeah, merely returning the action doesn't get it dispatched, like in generators. Good catch!

Comment (ntsekouras) [reply packages/edit-post/src/store/actions.js]: Removed it 😄

**Comment by jsnajdr** [comment]
The e2e tests are now green after the `META_BOXES_INITIALIZED`, which means we can merge? :tada:
